tools/gfx/debug-layer.cpp

```diff
@@ -0,0 +1,948 @@
+        return result;
+    returnComPtr(outCommandBuffer, outObject);
+    return result;
+}
+
+const ISwapchain::Desc& DebugSwapchain::getDesc()
+{
+    SLANG_GFX_API_FUNC;
+    desc = baseObject->getDesc();
+    desc.queue = queue.Ptr();
+    return desc;
+}
+
+Result DebugSwapchain::getImage(uint32_t index, ITextureResource** outResource)
+{
+    SLANG_GFX_API_FUNC;
+    maybeRebuildImageList();
+    if (index > (uint32_t)m_images.getCount())
+    {
+        GFX_DIAGNOSE_ERROR_FORMAT(
+            "`index`(%d) must not exceed total number of images (%d) in the swapchain.",
+            index,
+            (uint32_t)m_images.getCount());
+    }
+    returnComPtr(outResource, m_images[index]);
+    return SLANG_OK;
+}
+
+Result DebugSwapchain::present()
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->present();
+}
+
+int DebugSwapchain::acquireNextImage()
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->acquireNextImage();
+}
+
+Result DebugSwapchain::resize(uint32_t width, uint32_t height)
+{
+    SLANG_GFX_API_FUNC;
+    for (auto& image : m_images)
+    {
+        if (image->debugGetReferenceCount() != 1)
+        {
+            GFX_DIAGNOSE_ERROR("all swapchain images must be released before calling resize().");
+            return SLANG_FAIL;
+        }
+    }
+    m_images.clearAndDeallocate();
+    return baseObject->resize(width, height);
+}
+
+void DebugSwapchain::maybeRebuildImageList()
+{
+    SLANG_GFX_API_FUNC;
+    if (m_images.getCount() != 0)
+        return;
+    m_images.clearAndDeallocate();
+    for (uint32_t i = 0; i < baseObject->getDesc().imageCount; i++)
+    {
+        RefPtr<DebugTextureResource> image = new DebugTextureResource();
+        baseObject->getImage(i, image->baseObject.writeRef());
+        m_images.add(image);
+    }
+}
+
+slang::TypeLayoutReflection* DebugShaderObject::getElementTypeLayout()
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->getElementTypeLayout();
+}
+
+UInt DebugShaderObject::getEntryPointCount()
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->getEntryPointCount();
+}
+
+Result DebugShaderObject::getEntryPoint(UInt index, IShaderObject** entryPoint)
+{
+    SLANG_GFX_API_FUNC;
+    if (m_entryPoints.getCount() == 0)
+    {
+        for (UInt i = 0; i < getEntryPointCount(); i++)
+        {
+            RefPtr<DebugShaderObject> entryPointObj = new DebugShaderObject();
+            SLANG_RETURN_ON_FAIL(
+                baseObject->getEntryPoint(i, entryPointObj->baseObject.writeRef()));
+            m_entryPoints.add(entryPointObj);
+        }
+    }
+    if (index > (UInt)m_entryPoints.getCount())
+    {
+        GFX_DIAGNOSE_ERROR("`index` must not exceed `entryPointCount`.");
+        return SLANG_FAIL;
+    }
+    returnComPtr(entryPoint, m_entryPoints[index]);
+    return SLANG_OK;
+}
+
+Result DebugShaderObject::setData(ShaderOffset const& offset, void const* data, size_t size)
+{
+    SLANG_GFX_API_FUNC;
+    return baseObject->setData(offset, data, size);
```

Comment:
Obvious validation opportunity here is to calculate the ordinary data size for an object at creation time and validate that the range being set here is inside the object.